### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by omarcr** [pavlin-policar/fastTSNE#8]
Hey guys,
I receive the following error when trying to install via pip and anaconda:

Microsoft Windows [Version 10.0.17134.285]
(c) 2018 Microsoft Corporation. All rights reserved.

```
C:\Users\Omar>conda activate tensorflow

(tensorflow) C:\Users\Omar>pip install fasttsne
Collecting fasttsne
  Downloading https://files.pythonhosted.org/packages/32/98/ad7b5278e9fbc0c6f52a336a12440cc405ba620526e30f62a6ea5516829d/fastTSNE-0.2.6.tar.gz (439kB)
    100% |################################| 440kB 6.6MB/s
Requirement already satisfied: numpy>1.14 in c:\anaconda2\envs\tensorflow\lib\site-packages (from fasttsne) (1.15.0)
Collecting numba>=0.38.1 (from fasttsne)
  Downloading https://files.pythonhosted.org/packages/aa/ac/124fe0c1ad1ff7cbe0311352b545a142927583ffb44f57585d6a40b3f5a6/numba-0.39.0-cp35-cp35m-win_amd64.whl (1.6MB)
    100% |################################| 1.6MB 3.1MB/s
Requirement already satisfied: scikit-learn<0.19.99,>=0.19 in c:\anaconda2\envs\tensorflow\lib\site-packages (from fasttsne) (0.19.2)
Requirement already satisfied: scipy in c:\anaconda2\envs\tensorflow\lib\site-packages (from fasttsne) (1.1.0)
Collecting llvmlite>=0.24.0dev0 (from numba>=0.38.1->fasttsne)
  Downloading https://files.pythonhosted.org/packages/0b/7b/0ed27a8429d674ec82631f71d3896f23dd0caa87d5313bcb4535c58d7ab1/llvmlite-0.24.0-cp35-cp35m-win_amd64.whl (10.6MB)
    100% |################################| 10.6MB 3.3MB/s
Building wheels for collected packages: fasttsne
  Running setup.py bdist_wheel for fasttsne ... error
  Complete output from command C:\Anaconda2\envs\tensorflow\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\Omar\\AppData\\Local\\Temp\\pip-install-jq2cu5np\\fasttsne\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d C:\Users\Omar\AppData\Local\Temp\pip-wheel-0p5n70do --python-tag cp35:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.5
  creating build\lib.win-amd64-3.5\benchmarks
  copying benchmarks\benchmark_tsne.py -> build\lib.win-amd64-3.5\benchmarks
  copying benchmarks\__init__.py -> build\lib.win-amd64-3.5\benchmarks
  creating build\lib.win-amd64-3.5\fastTSNE
  copying fastTSNE\affinity.py -> build\lib.win-amd64-3.5\fastTSNE
  copying fastTSNE\callbacks.py -> build\lib.win-amd64-3.5\fastTSNE
  copying fastTSNE\metrics.py -> build\lib.win-amd64-3.5\fastTSNE
  copying fastTSNE\nearest_neighbors.py -> build\lib.win-amd64-3.5\fastTSNE
  copying fastTSNE\tsne.py -> build\lib.win-amd64-3.5\fastTSNE
  copying fastTSNE\__init__.py -> build\lib.win-amd64-3.5\fastTSNE
  creating build\lib.win-amd64-3.5\tests
  copying tests\test_correctness.py -> build\lib.win-amd64-3.5\tests
  copying tests\test_tsne.py -> build\lib.win-amd64-3.5\tests
  copying tests\__init__.py -> build\lib.win-amd64-3.5\tests
  creating build\lib.win-amd64-3.5\fastTSNE\pynndescent
  copying fastTSNE\pynndescent\distances.py -> build\lib.win-amd64-3.5\fastTSNE\pynndescent
  copying fastTSNE\pynndescent\pynndescent_.py -> build\lib.win-amd64-3.5\fastTSNE\pynndescent
  copying fastTSNE\pynndescent\rp_trees.py -> build\lib.win-amd64-3.5\fastTSNE\pynndescent
  copying fastTSNE\pynndescent\utils.py -> build\lib.win-amd64-3.5\fastTSNE\pynndescent
  copying fastTSNE\pynndescent\__init__.py -> build\lib.win-amd64-3.5\fastTSNE\pynndescent
  running build_ext
  building 'fastTSNE.quad_tree' extension
  error: Microsoft Visual C++ 14.0 is required. Get it with "Microsoft Visual C++ Build Tools": https://visualstudio.microsoft.com/downloads/

  ----------------------------------------
  Failed building wheel for fasttsne
  Running setup.py clean for fasttsne
Failed to build fasttsne
Installing collected packages: llvmlite, numba, fasttsne
  Running setup.py install for fasttsne ... error
    Complete output from command C:\Anaconda2\envs\tensorflow\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\Omar\\AppData\\Local\\Temp\\pip-install-jq2cu5np\\fasttsne\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\Omar\AppData\Local\Temp\pip-record-wft2hl9z\install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build\lib.win-amd64-3.5
    creating build\lib.win-amd64-3.5\benchmarks
    copying benchmarks\benchmark_tsne.py -> build\lib.win-amd64-3.5\benchmarks
    copying benchmarks\__init__.py -> build\lib.win-amd64-3.5\benchmarks
    creating build\lib.win-amd64-3.5\fastTSNE
    copying fastTSNE\affinity.py -> build\lib.win-amd64-3.5\fastTSNE
    copying fastTSNE\callbacks.py -> build\lib.win-amd64-3.5\fastTSNE
    copying fastTSNE\metrics.py -> build\lib.win-amd64-3.5\fastTSNE
    copying fastTSNE\nearest_neighbors.py -> build\lib.win-amd64-3.5\fastTSNE
    copying fastTSNE\tsne.py -> build\lib.win-amd64-3.5\fastTSNE
    copying fastTSNE\__init__.py -> build\lib.win-amd64-3.5\fastTSNE
    creating build\lib.win-amd64-3.5\tests
    copying tests\test_correctness.py -> build\lib.win-amd64-3.5\tests
    copying tests\test_tsne.py -> build\lib.win-amd64-3.5\tests
    copying tests\__init__.py -> build\lib.win-amd64-3.5\tests
    creating build\lib.win-amd64-3.5\fastTSNE\pynndescent
    copying fastTSNE\pynndescent\distances.py -> build\lib.win-amd64-3.5\fastTSNE\pynndescent
    copying fastTSNE\pynndescent\pynndescent_.py -> build\lib.win-amd64-3.5\fastTSNE\pynndescent
    copying fastTSNE\pynndescent\rp_trees.py -> build\lib.win-amd64-3.5\fastTSNE\pynndescent
    copying fastTSNE\pynndescent\utils.py -> build\lib.win-amd64-3.5\fastTSNE\pynndescent
    copying fastTSNE\pynndescent\__init__.py -> build\lib.win-amd64-3.5\fastTSNE\pynndescent
    running build_ext
    building 'fastTSNE.quad_tree' extension
    error: Microsoft Visual C++ 14.0 is required. Get it with "Microsoft Visual C++ Build Tools": https://visualstudio.microsoft.com/downloads/

    ----------------------------------------
Command "C:\Anaconda2\envs\tensorflow\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\Omar\\AppData\\Local\\Temp\\pip-install-jq2cu5np\\fasttsne\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\Omar\AppData\Local\Temp\pip-record-wft2hl9z\install-record.txt --single-version-externally-managed --compile" failed with error code 1 in C:\Users\Omar\AppData\Local\Temp\pip-install-jq2cu5np\fasttsne\
Cache entry deserialization failed, entry ignored
```

I have already installed visual c++ before.

**Comment by pavlin-policar** [comment]
Unfortunately, I have next to zero experience dealing with compiling cython extensions on Windows. I am currently waiting for the [conda feedstock repository](https://github.com/conda-forge/fasttsne-feedstock) to be set up and after that is up this package should be installable through conda.

Other than that, my best guess, looking at the error is that the compiler is not in your `PATH` so it's not found.

I'm sorry I couldn't be more helpful.

**Comment by pavlin-policar** [comment]
So the conda repository was set up a while ago, and I believe that conda uses it's own C++ compiler, so simply running

```
conda install fasttsne
```

should work. @omarcr, could you verify?

**Comment by pavlin-policar** [comment]
I am closing this due to lack of activity. If the issue persists, feel free to re-open.

**Comment by omarcr** [comment]
Hi Pavlin, thank you for notifying me:
I have tested the code. While I can confirm that the code works, the code is actually slower than the sklearn library:

Sklearn time on MNIST: 4.52 seconds
FasttSNE time on MNIST: 23.65 seconds

Please refer to the following code for testing:


```
# Authors: Fabian Pedregosa <fabian.pedregosa@inria.fr>
#          Olivier Grisel <olivier.grisel@ensta.org>
#          Mathieu Blondel <mathieu@mblondel.org>
#          Gael Varoquaux
# License: BSD 3 clause (C) INRIA 2011

print(__doc__)
from time import time

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import offsetbox
from sklearn import (manifold, datasets, decomposition, ensemble,
                     discriminant_analysis, random_projection)

digits = datasets.load_digits(n_class=6)
X = digits.data
y = digits.target
n_samples, n_features = X.shape
n_neighbors = 30

#import MulticoreTSNE as TSNE

from fastTSNE import TSNE
from examples import utils  # Utils for loading data and plotting embeddings

#----------------------------------------------------------------------
# Scale and visualize the embedding vectors

def plot_embedding(X, title=None):
    x_min, x_max = np.min(X, 0), np.max(X, 0)
    X = (X - x_min) / (x_max - x_min)

    plt.figure()
    ax = plt.subplot(111)
    for i in range(X.shape[0]):
        plt.text(X[i, 0], X[i, 1], str(y[i]),
                 color=plt.cm.Set1(y[i] / 10.),
                 fontdict={'weight': 'bold', 'size': 9})

    if hasattr(offsetbox, 'AnnotationBbox'):
        # only print thumbnails with matplotlib > 1.0
        shown_images = np.array([[1., 1.]])  # just something big
        for i in range(X.shape[0]):
            dist = np.sum((X[i] - shown_images) ** 2, 1)
            if np.min(dist) < 4e-3:
                # don't show points that are too close
                continue
            shown_images = np.r_[shown_images, [X[i]]]
            imagebox = offsetbox.AnnotationBbox(
                offsetbox.OffsetImage(digits.images[i], cmap=plt.cm.gray_r),
                X[i])
            ax.add_artist(imagebox)
    plt.xticks([]), plt.yticks([])
    if title is not None:
        plt.title(title)


#----------------------------------------------------------------------
# Plot images of the digits
n_img_per_row = 20
img = np.zeros((10 * n_img_per_row, 10 * n_img_per_row))
for i in range(n_img_per_row):
    ix = 10 * i + 1
    for j in range(n_img_per_row):
        iy = 10 * j + 1
        img[ix:ix + 8, iy:iy + 8] = X[i * n_img_per_row + j].reshape((8, 8))

plt.imshow(img, cmap=plt.cm.binary)
plt.xticks([])
plt.yticks([])
plt.title('A selection from the 64-dimensional digits dataset')


#----------------------------------------------------------------------
# t-SNE embedding of the digits dataset
print("Computing t-SNE embedding")

# sklearn t-SNE embedding of the digits dataset
print("Computing t-SNE embedding")
# tsne = manifold.TSNE(n_components=2, init='pca', random_state=0)
# t0 = time()
# X_tsne = tsne.fit_transform(X)
#
# plot_embedding(X_tsne,
#                "t-SNE embedding of the digits (time %.2fs)" %
#                (time() - t0))
#
# plt.show()


#fast TSNE

tsne = TSNE(n_components=2, initialization='pca', n_jobs=-1)
t0 = time()
embedding  = tsne.fit(X)
#fig = plt.figure(figsize=(13, 8))

#utils.plot(embedding, y, ax=fig.gca())

plot_embedding(embedding,
               "t-SNE embedding of the digits (time %.2fs)" %
               (time() - t0))
plt.show()
```

**Comment by pavlin-policar** [comment]
Hey, glad to hear it works now!

> While I can confirm that the code works, the code is actually slower than the sklearn library.

That is very strange indeed. I created a completely new conda environment and installed fastTSNE. I trimmed down your code to the bare minimum and unfortunately, I am unable to replicate your results.

```
from time import time

from fastTSNE import TSNE
from sklearn import manifold, datasets

digits = datasets.load_digits(n_class=6)
X, y = digits.data, digits.target

# sklearn t-SNE embedding of the digits dataset
tsne = manifold.TSNE(n_components=2, init='pca', random_state=0)
t0 = time()
X_tsne = tsne.fit_transform(X)
print('scikit-learn: %.2f' % (time() - t0))

# fast TSNE
tsne = TSNE(n_components=2, initialization='pca', n_jobs=4)
t0 = time()
embedding = tsne.fit(X)
print('fasttsne: %.2f' % (time() - t0))
```
and I get the following output
```
scikit-learn: 12.80
fasttsne: 2.00
```

The exact times are sure to be different since we are running different hardware, but might it be possible that you mixed your numbers up and that it was fastTSNE which ran in 4.5 seconds and sklearn that ran in 24 seconds? The speed-up in my run was 6x, which is the same as the speed-up you observed.

Honestly, the only other thing that comes to mind that would explain these results is if you were running fastTSNE with the parameter settings meant for large data sets, like this:
```
tsne = TSNE(n_components=2, initialization='pca', n_jobs=4,
            negative_gradient_method='fft', neighbors='approx')
t0 = time()
embedding = tsne.fit(X)
print('fasttsne: %.2f' % (time() - t0))
```
which results in
```
fasttsne: 58.87
```

Could you run the snippet I provided (it does the same as yours, without the plotting) and paste the output of that and verify that scikit-learn is still faster than fastTSNE? If so, please provide your CPU model, operating system, python, numpy versions which will be useful for debugging this strange behaviour.

**Comment by omarcr** [comment]
Hey Pavlin, thanks for the test.
I have also started a blank environment to test this further.

Here is the output of the installation:

```

C:\Users\Omar>conda activate test2

(test2) C:\Users\Omar>conda install -c conda-forge fasttsne
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.9
  latest version: 4.5.11

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: C:\Anaconda2\envs\test2

  added / updated specs:
    - fasttsne


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    scipy-1.1.0                |   py35h4f6bf74_1        13.5 MB
    scikit-learn-0.19.2        |   py35heebcf9a_0         4.7 MB
    numpy-base-1.15.2          |   py35h8128ebf_0         3.9 MB
    intel-openmp-2019.0        |              118         1.7 MB
    mkl-2019.0                 |              118       178.1 MB
    numpy-1.15.2               |   py35ha559c80_0          48 KB
    ------------------------------------------------------------
                                           Total:       201.8 MB

The following NEW packages will be INSTALLED:

    blas:         1.0-mkl
    fasttsne:     0.2.6-py35h452e1ab_0  conda-forge
    icc_rt:       2017.0.4-h97af966_0
    intel-openmp: 2019.0-118
    llvmlite:     0.24.0-py35_1         conda-forge
    mkl:          2019.0-118
    mkl_fft:      1.0.6-py35_0          conda-forge
    mkl_random:   1.0.1-py35_0          conda-forge
    numba:        0.39.0-py35h830ac7b_0
    numpy:        1.15.2-py35ha559c80_0
    numpy-base:   1.15.2-py35h8128ebf_0
    scikit-learn: 0.19.2-py35heebcf9a_0
    scipy:        1.1.0-py35h4f6bf74_1

The following packages will be UPDATED:

    certifi:      2018.8.24-py35_0      conda-forge --> 2018.8.24-py35_1001 conda-forge

Proceed ([y]/n)?


Downloading and Extracting Packages
scipy-1.1.0          | 13.5 MB | ################################################################################################################################################################################################################################################################################################################ | 100%
scikit-learn-0.19.2  |  4.7 MB | ################################################################################################################################################################################################################################################################################################################ | 100%
numpy-base-1.15.2    |  3.9 MB | ################################################################################################################################################################################################################################################################################################################ | 100%
intel-openmp-2019.0  |  1.7 MB | ################################################################################################################################################################################################################################################################################################################ | 100%
mkl-2019.0           | 178.1 MB | ############################################################################################################################################################################################################################################################################################################### | 100%
numpy-1.15.2         |   48 KB | ################################################################################################################################################################################################################################################################################################################ | 100%
Preparing transaction: done
Verifying transaction: done
Executing transaction: done

(test2) C:\Users\Omar>
(test2) C:\Users\Omar>
(test2) C:\Users\Omar>


```
Output of your minimum example program:

```
C:\Anaconda2\envs\test2\python.exe C:/Users/Omar/PycharmProjects/test_fastsne/testing.py
scikit-learn: 15.09
nn 0.16428399085998535
optmization 23.613564491271973
fasttsne: 23.78

Process finished with exit code 0

```


This is a very strange behavior indeed.
I am not sure what is interfering with the speed, but is definitely inside the optimization routine (as seen above) 

I hope more people can test this out.

**Comment by pavlin-policar** [comment]
This is very strange indeed. I have access to a windows machine and I will try to replicate this there.

Two things stand out to me:
1. conda appears to be installing version 0.2.6, the latest version is 0.2.7 and should be available on conda-forge, so this is strange. In version 0.2.6, the `n_jobs=-1` parameter wasn't working as expected and was likely using a single core. However, in my snippet, I set `n_jobs=4`, which should have worked fine in that version, so this can't be the culprit.
2. In your output, you have the following lines:
```
nn 0.16428399085998535
optmization 23.613564491271973
```
I remember having something similar for debugging a long time ago, and I never added this to git. It must have somehow snuck its way into the pip package. In any case, this indicates that you're using an old version of the package (as supported by the first point). If the print statements somehow made it into the package, I'm not sure what else might be there. Maybe I accidentaly included something that hurt performance...

Could you please make sure you're using the latest version of the package i.e. 0.2.7. I am not sure why conda didn't download that automatically for you... I don't know if this is responsible for the speed issues, but it is a good starting point.

I'll try to replicate this on Windows and get back to you.

**Comment by pavlin-policar** [comment]
Scratch that. I installed v0.2.6 on my linux machine and got similar results to yours. Apparently, there is something very wrong with v0.2.6. I am farily sure that the git version is fine, but something must have snuck into the pypi package. I'll either remove the version altogether or update the package on pypi. v0.2.7 fixed a couple important bugs, so it might be smarter to remove 0.2.6 altogether.

Could you update and check if this fixes the problem?

**Comment by pavlin-policar** [comment]
Ok, I see what the problem is. As of 1.10.2018, conda stopped supporting Python 3.5 ([announcement](https://conda-forge.org/docs/news_announce.html#pinned-announcements)). This means that any conda package was no longer being built for 3.5.

I had pushed v0.2.7 to conda only a few days ago, so the only available 0.2.7 package is for Python 3.6. Unfortunately, this means that only the broken 0.2.6 package is available for Python 3.5.

> Using pip assumes a linux environment

Yes, that's why I recommend installing via conda. You can use pip if you have a C compiler available, or you can let conda precompile the packages for you.

Fortunately, the solution is really simple - just update your python version. This can be done easily with conda

```
conda install python==3.6
```

I don't know what I can do about this, I'll see if I can remove the broken package from conda altogether.

**Comment by omarcr** [comment]
I have updated to 3.6

This is the output of the minimum example that you posted:

C:\Anaconda2\envs\tensorflow\python.exe C:/Users/Omar/PycharmProjects/test_fastsne/testing.py
scikit-learn: 15.24
fasttsne: 24.08

Process finished with exit code 0

There are no "optimization" output but the algorithm is still slower thank sklearn library.

Omar

**Comment by pavlin-policar** [comment]
I believe this was fixed in #19 and should be fixed in all versions of the package on both PyPi and conda-forge. If not, please feel free to re-open.
